packages/xod-project/src/patch.js

```diff
@@ -1354,3 +1355,140 @@ export const validateAbstractPatch = def(
     Either.of
   )
 );
+
+export const getPatchSignature = patch =>
+  R.compose(
+    R.map(
+      R.compose(
+        R.fromPairs,
+        R.map(R.converge(R.pair, [Pin.getPinOrder, Pin.getPinType]))
+      )
+    ),
+    R.groupBy(Pin.getPinDirection),
+    listPins
+  )(patch);
+
+// checks that a concrete patch conforms to concrete patch specification
+export const validateAbstractPatchImplementation = def(
```

Comment:
The “implementation” term is tied to “a portion of C++” in XOD… Would it be correct to name this `validateAbstractPatchSpecialization`? Or even `checkSpecializationMatchesAbstraction`.